### PR TITLE
GH Pages: HTML minifier and prod build for assets #291

### PR DESCRIPTION
**Pull request by NastaLeo**
1. HTML is compressing via plugin called html-minifier (eleventy.js config file).
2. CSS optimization is implemented with using clean-css and cssnano in npm scripts.
3. Minification of JS bundles was realized by specifying  production mode in npm scripts.

### CODE REVIEW

**Comment by NastaLeo** [comment]
I have read the CLA Document and I hereby sign the CLA

File: .eleventy.js [hunk 1]
```diff
@@ -1,38 +1,53 @@
-const { isDev } = require('./pages/views/_data/env')
+const { isDev } = require("./pages/views/_data/env");
+const htmlmin = require("html-minifier");
 
-module.exports = config => {
+module.exports = (config) => {
   config.addPassthroughCopy({
-    'pages/static/assets': 'assets',
-    'pages/static/tools': '.',
+    "pages/static/assets": "assets",
```
Comment (ala-n) [inline .eleventy.js:6]: Is it compliant with our editor config rules?
If I'm not mistaken we should use single quotes for JS and TS

File: package.json [hunk 2]
```diff
@@ -87,26 +93,33 @@
     "@typescript-eslint/parser": "^4.28.3",
     "@webcomponents/custom-elements": "^1.4.3",
     "@webcomponents/webcomponents-platform": "^1.0.1",
+    "autoprefixer": "^10.3.1",
```
Comment (ala-n) [inline package.json:96]: Do we really need autoprefixer?

File: package.json [hunk 3]
```diff
@@ -87,26 +93,33 @@
...
     "browser-sync": "^2.27.4",
     "chokidar-cli": "^2.1.0",
+    "clean-css-cli": "^5.3.0",
     "concurrently": "^6.2.0",
     "copyfiles": "^2.4.1",
+    "cssnano": "^5.0.7",
```
Comment (ala-n) [inline package.json:102]: Do we need both cssnano + clean-css, it sounds like one solution will be enogh.

File: package.json [hunk 4]
```diff
@@ -87,26 +93,33 @@
...
-    "postcss": "^8.3.5",
+    "postcss": "^8.3.6",
+    "postcss-cli": "^8.3.1",
     "promise-polyfill": "^8.2.0",
     "rimraf": "^3.0.2",
     "semantic-release": "^17.4.4",
     "smoothscroll-polyfill": "^0.4.4",
+    "style-loader": "^3.2.1",
     "stylelint": "^13.13.1",
+    "terser-webpack-plugin": "^5.1.4",
```
Comment (ala-n) [inline package.json:122]: Do we need?

File: package.json [hunk 5]
```diff
@@ -87,26 +93,33 @@
...
     "less": "^4.1.1",
     "lintspaces-cli": "^0.7.1",
-    "postcss": "^8.3.5",
+    "postcss": "^8.3.6",
+    "postcss-cli": "^8.3.1",
     "promise-polyfill": "^8.2.0",
     "rimraf": "^3.0.2",
     "semantic-release": "^17.4.4",
     "smoothscroll-polyfill": "^0.4.4",
+    "style-loader": "^3.2.1",
```
Comment (ala-n) [inline package.json:120]: Do we need?

File: package.json [hunk 6]
```diff
@@ -48,10 +48,16 @@
...
+    "minify:cssnano": "postcss --use cssnano -o pages/dist/bundles/localdev.css pages/dist/bundles/localdev.css",
     "build-pages": "concurrently \"npm run build-pages:ts\" \"npm run build-pages:less\" && npm run build-pages:eleventy",
+    "build-pages:prod": "concurrently \"npm run build-pages:prod:ts\" \"npm run build-pages:prod:less\" && npm run build-pages:prod:eleventy",
     "build-pages:ts": "cd pages && webpack",
+    "build-pages:prod:ts": "cd pages && webpack --env=production --mode=production",
     "build-pages:less": "lessc pages/src/localdev.less pages/dist/bundles/localdev.css --source-map=pages/dist/bundles/localdev.css.map",
-    "build-pages:eleventy": "npx @11ty/eleventy -- --env=production",
+    "build-pages:prod:less": "npm run build-pages:less && npm run minify:css && npm run minify:cssnano",
+    "build-pages:eleventy": "npx @11ty/eleventy",
+    "build-pages:prod:eleventy": "npx @11ty/eleventy -- --env=production",
```
Comment (dshovchko) [inline package.json:60]: It seems to me that more naturally use by default prod build and custom build for dev (`dev` instead `prod` in script names).
Аnd order in script names... I guess the name of the environment should be moved to the end of the name.
`build-pages:eleventy:prod` instead `build-pages:prod:eleventy`.
What do you think?

File: .eleventy.js [hunk 7]
```diff
@@ -1,38 +1,53 @@
...
+      "pages/dist/bundles/*.js",
+      "pages/dist/bundles/*.css",
+      "pages/dist/bundles/*.map",
     ],
     open: isDev,
   });
 
+  config.addTransform("htmlmin", function (content, outputPath) {
+    if (outputPath && outputPath.endsWith(".html") && !isDev) {
+      return require(`html-minifier`).minify(content, {
```
Comment (YanaBr) [inline .eleventy.js:29]: As I see you require html-minifier on 2-nd line and put it in `htmlmin` variable, but here you require it again. Is it correct or we can use `htmlmin` variable from the 2-nd line?

Comment (NastaLeo) [reply .eleventy.js:29]: Thank you for perfect note.  I've considered your remark and now use htmlmin variable.

File: package.json (hunk 5 above)
Comment (NastaLeo) [reply package.json:120]: Sorry, I didn't remove unused dependencies from package.json.

File: .eleventy.js (hunk 1 above)
Comment (NastaLeo) [reply .eleventy.js:6]: Thank you for good note. It's my mistake and it will be fixed.

File: package.json (hunk 6 above)
Comment (NastaLeo) [reply package.json:60]: It's a logical proposal. I will take it into account and will refactor scripts in package.json.

**Comment by NastaLeo** [comment]
It's a new commit based on the above-mentioned suggestions

File: package.json [hunk 8]
```diff
@@ -48,12 +48,17 @@
...
+    "minify:css": "cleancss -o pages/dist/bundles/localdev.css pages/dist/bundles/localdev.css",
+    "build-pages:dev": "concurrently \"npm run build-pages:ts:dev\" \"npm run build-pages:less:dev\" && npm run build-pages:eleventy:dev",
     "build-pages": "concurrently \"npm run build-pages:ts\" \"npm run build-pages:less\" && npm run build-pages:eleventy",
-    "build-pages:ts": "cd pages && webpack",
-    "build-pages:less": "lessc pages/src/localdev.less pages/dist/bundles/localdev.css --source-map=pages/dist/bundles/localdev.css.map",
+    "build-pages:ts:dev": "cd pages && webpack",
+    "build-pages:ts": "cd pages && webpack --env=production --mode=production",
+    "build-pages:less:dev": "lessc pages/src/localdev.less pages/dist/bundles/localdev.css --source-map=pages/dist/bundles/localdev.css.map",
+    "build-pages:less": "npm run build-pages:less:dev && npm run minify:css",
+    "build-pages:eleventy:dev": "npx @11ty/eleventy",
```
Comment (abarmina) [inline package.json:58]: Do we not need to pass `env= development` as an attribute here?

Comment (NastaLeo) [reply package.json]: I've changed 'build-pages' scripts

**Review by ala-n — APPROVED** [review]

**Review by dshovchko — APPROVED** [review]

File: package.json [hunk 9]
```diff
@@ -48,12 +48,13 @@
     "build:less:cpy": "copyfiles \"src/**/*.less\" \".\" --up 1 --all",
     "build:less:css": "foreach -x \"lessc #{path} #{dir}/#{name}.css\" -i \"**/*.mixin.less\" -g \"modules/{*,*/*,{draft,esl-utils,esl-forms}/*/*}.less\"",
     "build:docs": "copyfiles \"src/**/*.md\" \".\" --up 1 --all",
+    "minify:css": "cleancss -o pages/dist/bundles/localdev.css pages/dist/bundles/localdev.css",
     "build-pages": "concurrently \"npm run build-pages:ts\" \"npm run build-pages:less\" && npm run build-pages:eleventy",
-    "build-pages:ts": "cd pages && webpack",
-    "build-pages:less": "lessc pages/src/localdev.less pages/dist/bundles/localdev.css --source-map=pages/dist/bundles/localdev.css.map",
+    "build-pages:ts": "cd pages && webpack --mode=production",
+    "build-pages:less": "lessc pages/src/localdev.less pages/dist/bundles/localdev.css --source-map=pages/dist/bundles/localdev.css.map && npm run minify:css",
```
Comment (abarmina) [inline package.json:54]: it looks great, but now we have repeating long code `lessc pages/src/localdev.less pages/dist/bundles/localdev.css --source-map=pages/dist/bundles/localdev.css.map` on two lines 54 and 57. maybe we can put it into a separate script, and use it in script build-pages `\"npm run build-pages:less && npm run minify:css\"`. it's like a suggestion

**Review by abarmina — APPROVED** [review]

**Review by julia-murashko — APPROVED** [review]

**Review by YanaBr — APPROVED** [review]

**Pull request merged**